In [1]:
# Data Analysis and Processing
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Model Related
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier

# Feature Selection
from sklearn.feature_selection import SelectFromModel

In [2]:
df = pd.read_csv("Telco-Customer-Churn.csv")

In [3]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

## Data Processing

In [4]:
# Dropping the customer ID column
df = df.drop('customerID', axis = 1)

In [5]:
# Converting TotalCharges to numeric
df[df['TotalCharges'] == " "] = 0
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'])

In [6]:
# Checking categorical variable values
cat_col = df.drop(['MonthlyCharges', 'TotalCharges', 'tenure'], axis = 1).columns

In [7]:
# Dropping columns with all 0
df = df.drop(df[df['gender'] == 0].index)

In [8]:
# Converting 'Yes' and 'No' to 1 and 0
binary_col = ["Partner", 
              "Dependents", 
              "PhoneService",
              "MultipleLines",
              "OnlineSecurity", 
              "OnlineBackup", 
              "DeviceProtection", 
              "TechSupport",
              "StreamingTV",
              "StreamingMovies",
              "PaperlessBilling",
              "Churn"]

df[binary_col] = df[binary_col].replace(to_replace=['No', 'Yes', 'No phone service', 'No internet service'], 
                                        value=[0, 1, 0, 0])

In [9]:
# Converting remaining categorical variables (gender, InternetService, Contract, PaymentMethod)
# Scale variable if ordinal
# Use dummy variables if nominal

nominal_col = ["gender",
               "InternetService",
               "Contract",
               "PaymentMethod"]

# Creating the dummy columns
gender_dummy = pd.get_dummies(df['gender'])
InternetService_dummy = pd.get_dummies(df['InternetService'])
Contract_dummy = pd.get_dummies(df['Contract'])
PaymentMethod_dummy = pd.get_dummies(df['PaymentMethod'])

# Dropping the original columns
df = df.drop(nominal_col, axis = 1)

# Concatenating dummy columns
df = pd.concat([df, gender_dummy], axis=1)
df = pd.concat([df, InternetService_dummy], axis=1)
df = pd.concat([df, Contract_dummy], axis=1)
df = pd.concat([df, PaymentMethod_dummy], axis=1)

In [10]:
# Saving the column names
col_names = df.columns

# Now we scale the remaining values
scaler = MinMaxScaler()
scaled = scaler.fit_transform(df)
df = pd.DataFrame(scaled, columns=col_names)

In [11]:
df.head()

SeniorCitizen  Partner  Dependents    tenure  PhoneService  MultipleLines  \
0            0.0      1.0         0.0  0.000000           0.0            0.0   
1            0.0      0.0         0.0  0.464789           1.0            0.0   
2            0.0      0.0         0.0  0.014085           1.0            0.0   
3            0.0      0.0         0.0  0.619718           0.0            0.0   
4            0.0      0.0         0.0  0.014085           1.0            0.0   

   OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  ...  DSL  \
0             0.0           1.0               0.0          0.0  ...  1.0   
1             1.0           0.0               1.0          0.0  ...  1.0   
2             1.0           1.0               0.0          0.0  ...  1.0   
3             1.0           0.0               1.0          1.0  ...  1.0   
4             0.0           0.0               0.0          0.0  ...  0.0   

   Fiber optic   No  Month-to-month  One year  Two year  \
0          0.0  0.0             1.0       0.0       0.0   
1          0.0  0.0             0.0       1.0       0.0   
2          0.0  0.0             1.0       0.0       0.0   
3          0.0  0.0             0.0       1.0       0.0   
4          1.0  0.0             1.0       0.0       0.0   

   Bank transfer (automatic)  Credit card (automatic)  Electronic check  \
0                        0.0                      0.0               1.0   
1                        0.0                      0.0               0.0   
2                        0.0                      0.0               0.0   
3                        1.0                      0.0               0.0   
4                        0.0                      0.0               1.0   

   Mailed check  
0           0.0  
1           1.0  
2           1.0  
3           0.0  
4           0.0  

[5 rows x 28 columns]

## Creating a train test split for the model

In [12]:
X = df.iloc[:, df.columns != 'Churn']
y = df['Churn']
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.75, random_state = 101)

## Random Forest

In [13]:
# Defining the model
clf = RandomForestClassifier(n_estimators=10000, random_state=0, n_jobs=-1)
# Training the model
clf.fit(X_train, y_train)
# Making predictions
preds = clf.predict(X_test)
# Calculating accuracy
accuracy_score(y_test, preds)

0.7957906712172924

### Using all features, we get an accuracy of 79%

## Feature selection - Gini Impurity

In [14]:
features = list(zip(X_train.columns, clf.feature_importances_))
features.sort(key=lambda x: x[1], reverse=True)

for feature in features:
    print (feature)

('TotalCharges', 0.1806962902284617)
('MonthlyCharges', 0.1667634823111917)
('tenure', 0.15736480265397912)
('Month-to-month', 0.0668568747353479)
('Fiber optic', 0.03766942389972935)
('Electronic check', 0.03446236769350899)
('PaperlessBilling', 0.02532259711338781)
('Two year', 0.02252892316079394)
('Partner', 0.0224200003686668)
('OnlineSecurity', 0.021848118932735718)
('SeniorCitizen', 0.02103480928798737)
('OnlineBackup', 0.020649240539092083)
('TechSupport', 0.02026410520895087)
('Dependents', 0.019100385706117183)
('MultipleLines', 0.018409057657758656)
('DeviceProtection', 0.01799316854899335)
('StreamingMovies', 0.017500701658277468)
('StreamingTV', 0.017304470489468506)
('Male', 0.017204431974375396)
('Female', 0.017195389583280683)
('No', 0.01447058128154402)
('Credit card (automatic)', 0.01188497855781574)
('Bank transfer (automatic)', 0.011827593477151017)
('One year', 0.01117983438671837)
('Mailed check', 0.01103812760553605)
('DSL', 0.010550988494905893)
('PhoneService',

In [15]:
# Selecting features above 0.05
sfm = SelectFromModel(clf, threshold=0.05)

In [16]:
sfm.fit(X_train, y_train)

SelectFromModel(estimator=RandomForestClassifier(bootstrap=True,
                                                 class_weight=None,
                                                 criterion='gini',
                                                 max_depth=None,
                                                 max_features='auto',
                                                 max_leaf_nodes=None,
                                                 min_impurity_decrease=0.0,
                                                 min_impurity_split=None,
                                                 min_samples_leaf=1,
                                                 min_samples_split=2,
                                                 min_weight_fraction_leaf=0.0,
                                                 n_estimators=10000, n_jobs=-1,
                                                 oob_score=False,
                                                 random_state=0, verbose=0,
     

In [17]:
X_train.columns[sfm.get_support()]

Index(['tenure', 'MonthlyCharges', 'TotalCharges', 'Month-to-month'], dtype='object')

In [18]:
X_train_sfm = sfm.transform(X_train)
X_test_sfm = sfm.transform(X_test)

In [19]:
# Defining the model
clf = RandomForestClassifier(n_estimators=10000, random_state=0, n_jobs=-1)
# Training the model
clf.fit(X_train_sfm, y_train)
# Making predictions
preds = clf.predict(X_test_sfm)
# Calculating accuracy
accuracy_score(y_test, preds)

0.7718998862343572

### By reducing features to 4, we get an accuracy of 77%